services/src/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/utils/OfFlowPresenceVerifier.java

```diff
@@ -0,0 +1,183 @@
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.floodlight.utils;
+
+import lombok.Getter;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import org.projectfloodlight.openflow.protocol.OFFlowMod;
+import org.projectfloodlight.openflow.protocol.OFFlowStatsEntry;
+import org.projectfloodlight.openflow.protocol.action.OFAction;
+import org.projectfloodlight.openflow.protocol.instruction.OFInstruction;
+import org.projectfloodlight.openflow.protocol.instruction.OFInstructionWriteActions;
+import org.projectfloodlight.openflow.types.TableId;
+import org.projectfloodlight.openflow.types.U64;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class OfFlowPresenceVerifier {
+    private final Map<FlowLookupKey, List<OFFlowMod>> expected = new HashMap<>();
+
+    @Getter
+    private final CompletableFuture<OfFlowPresenceVerifier> finish = new CompletableFuture<>();
+
+    public OfFlowPresenceVerifier(OfFlowDumpProducer dumpProducer, List<OFFlowMod> expectedFlows) {
+        for (OFFlowMod entry : expectedFlows) {
+            FlowLookupKey key = new FlowLookupKey(entry.getTableId(), entry.getCookie());
+            log.debug("Schedule expect flow - {} - {}", key, entry);
+            expected.computeIfAbsent(key, ignore -> new ArrayList<>())
+                    .add(entry);
+        }
+
+        List<CompletableFuture<Void>> pending = new ArrayList<>();
+        for (CompletableFuture<List<OFFlowStatsEntry>> tableDump : dumpProducer.getTableRequests()) {
+            pending.add(tableDump.thenAccept(this::handleTableFlowStats));
+        }
+
+        CompletableFuture.allOf(pending.toArray(new CompletableFuture<?>[0]))
+                .whenComplete((ignore, error) -> handleComplete(error));
+    }
+
+    /**
+     * Expose "missing" OF flows.
+     */
+    public List<OFFlowMod> getMissing() {
+        return expected.values().stream()
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+    }
+
+    private void handleTableFlowStats(List<OFFlowStatsEntry> tableEntries) {
+        for (OFFlowStatsEntry entry : tableEntries) {
+            verifyTableEntry(entry);
+        }
+    }
+
+    private void handleComplete(Throwable error) {
```

Comment:
I think by the purpose this method is supposed to be named "handleError", right?

Reply:
Not really, it handles task completion (according to it's name) and the main part of it - check is it completed successfully or not, that's why it receive "error" argument.